src/main/java/org/example/bookstore/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,136 @@
+import org.example.bookstore.exception.EmptyCartException;
+import org.example.bookstore.exception.EntityNotFoundException;
+import org.example.bookstore.mapper.OrderItemMapper;
+import org.example.bookstore.mapper.OrderMapper;
+import org.example.bookstore.model.Book;
+import org.example.bookstore.model.CartItem;
+import org.example.bookstore.model.Order;
+import org.example.bookstore.model.OrderItem;
+import org.example.bookstore.model.User;
+import org.example.bookstore.repository.CartItemRepository;
+import org.example.bookstore.repository.OrderItemRepository;
+import org.example.bookstore.repository.OrderRepository;
+import org.example.bookstore.service.OrderService;
+import org.example.bookstore.service.ShoppingCartService;
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+
+    private final CartItemRepository cartItemRepository;
+    private final OrderRepository orderRepository;
+    private final OrderItemRepository orderItemRepository;
+    private final ShoppingCartService shoppingCartService;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+
+    @Override
+    public Set<OrderDto> getAll(Long userId) {
+        return orderRepository.findByCustomerIdWithItems(userId)
+                .stream()
+                .map(orderMapper::toDto)
+                .collect(Collectors.toSet());
+    }
+
+    @Transactional
+    @Override
+    public OrderDto createOrder(CreateOrderRequestDto orderRequest, Long userId) {
+        Set<CartItem> cartItems = cartItemRepository.findByCartId(userId);
+        if (cartItems.isEmpty()) {
+            throw new EmptyCartException("Cannot create order with an empty cart");
+        }
+        Order newOrder = createAndSaveOrder(orderRequest, userId, cartItems);
+        Set<OrderItem> orderItems = createAndSaveOrderItems(cartItems, newOrder);
+        newOrder.setOrderItems(orderItems);
+        shoppingCartService.clearShoppingCart(userId);
+        return orderMapper.toDto(newOrder);
+    }
+
+    @Override
+    public OrderDto updateOrderStatus(Long orderId, UpdateOrderStatusRequestDto requestDto) {
+        Order order =
+                orderRepository.findByOrderIdWithItems(orderId)
+                        .orElseThrow(() -> new EntityNotFoundException("Order not found"));
+        order.setStatus(requestDto.status());
+        orderRepository.save(order);
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    public Set<OrderItemDto> getOrderItems(Long orderId, Long userId) {
+        Order order =
+                orderRepository.findByOrderIdWithItems(orderId)
+                        .orElseThrow(() -> new EntityNotFoundException("Order not found"));
+        checkIfOrderBelongsToUser(order, userId);
```

Comment:
Is it possible that order you obtained by USER ID and order id can not belong to user?

Reply:
And if an order doesn`t exist at all I throw a NotFoundException and if it exists but it not belong to the current authorized user I throw an AccesDeniedException or it`s better to add filtering by user id on the db side and return a not found exception in this case?

Reply:
Discussed on Q&A and approved!